### PR TITLE
Append Exment Provider. 

### PR DESCRIPTION
**Pull request by hirossyi73**
Exment is open source web database for managing information assets on the Web.  
For Web Database, SFA, CRM, Business improvement, ...  

Please read them.
<a href="https://github.com/exceedone/exment" target="_blank">GitHub</a>  
<a href="https://exment.net/docs/#/">Manual</a>  
<a href="https://exment.net/docs/#/api">How to config API</a>  
<a href="https://exment.net/reference/webapi.html">API Reference</a>

### CODE REVIEW

File: src/Exment/NotConfigExmentUrlException.php [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace SocialiteProviders\Exment;
+
+class NotConfigExmentUrlException extends \Exception
```
Comment (lucasmichot) [inline src/Exment/NotConfigExmentUrlException.php:5]: This can be removed

File: src/Exment/Provider.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        return (new User())->setRaw($user)->map([
+            'id'   => $user['id'], 'nickname' => $user['value']['user_code'],
+            'name' => $user['value']['user_name'], 'email' => $user['value']['email'], 'avatar' => null,
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenFields($code)
```
Comment (lucasmichot) [inline src/Exment/Provider.php:83]: This can be removed

File: src/Exment/Provider.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        return array_merge(parent::getTokenFields($code), [
+            'grant_type' => 'authorization_code',
+        ]);
+    }
+
+    // Custom functions ----------------------------------------------
```
Comment (lucasmichot) [inline src/Exment/Provider.php:90]: This can be removed

File: src/Exment/Provider.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    /**
+     * Get Exment base uri.
+     *
+     * @return string
+     */
+    protected function getBaseUri(): string
+    {
+        $exment_url = $this->getConfig('exment_uri');
+        if (is_null($exment_url)) {
+            throw new NotConfigExmentUrlException();
```
Comment (lucasmichot) [inline src/Exment/Provider.php:100]: I would just throw an InvalidArgumentException with a custom message

File: src/Exment/Provider.php [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        return $exment_url;
+    }
+
+    /**
+     * Join UrlPath.
+     */
+    protected function urlJoin(...$pass_array)
```
Comment (lucasmichot) [inline src/Exment/Provider.php:109]: `urlJoin` is way overkill IMO, I would drop it
Can you just concatenate the urls

**Review by lucasmichot — CHANGES_REQUESTED** [review]
Some changes required

Also can you add an entry in `monorepo-builder.yml`

File: src/Exment/Provider.php (hunk 4 above)
Comment (hirossyi73) [reply src/Exment/Provider.php]: Changed to InvalidArgumentException.

File: src/Exment/Provider.php (hunk 5 above)
Comment (hirossyi73) [reply src/Exment/Provider.php]: Removed urlJoin, and join directly url. Ex. 

```

    /**
     * {@inheritdoc}
     */
    protected function getAuthUrl($state)
    {
        return $this->buildAuthUrlFromBase(
            $this->getBaseUri().'/oauth/authorize',
            $state
        );
    }
```

File: src/Exment/Provider.php (hunk 3 above)
Comment (hirossyi73) [reply src/Exment/Provider.php]: Removed

File: src/Exment/Provider.php (hunk 2 above)
Comment (hirossyi73) [reply src/Exment/Provider.php]: Removed.

File: src/Exment/NotConfigExmentUrlException.php (hunk 1 above)
Comment (hirossyi73) [reply src/Exment/NotConfigExmentUrlException.php]: Removed.

**Comment by hirossyi73** [comment]
Thanks for reviewing!
I changed, Please chack again.

File: src/Exment/Provider.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    protected $scopeSeparator = ' ';
+
+    /**
+     * The scopes being requested.
+     *
+     * @var array
+     */
+    protected $scopes = ['me'];
+
+    public static function additionalConfigKeys()
```
Comment (lucasmichot) [inline src/Exment/Provider.php:28]: Missing docblocks

**Review by lucasmichot — CHANGES_REQUESTED** [review]
Some other changes

File: src/Exment/Provider.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * Get Exment base uri.
+     *
+     * @return string
+     */
+    protected function getBaseUri(): string
+    {
+        $exment_url = $this->getConfig('exment_uri');
+        if (is_null($exment_url)) {
+            throw new \InvalidArgumentException('Please config Exment URI.');
```
[the remark this answers is not in the document]
Comment (hirossyi73) [reply src/Exment/Provider.php:89]: Changed.

File: src/Exment/Provider.php (hunk 6 above)
Comment (hirossyi73) [reply src/Exment/Provider.php:28]: Changed.

File: src/Exment/Provider.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+     * Get Exment base uri.
+     *
+     * @return string
+     */
+    protected function getBaseUri(): string
+    {
+        $exment_url = $this->getConfig('exment_uri');
```
[the remark this answers is not in the document]
Comment (hirossyi73) [reply src/Exment/Provider.php]: Changed to camel case.

**Comment by lucasmichot** [comment]
Some extra minor adjustments.

**Review by lucasmichot — APPROVED** [review]
GTM @atymic

**Review by atymic — APPROVED** [review]

**Pull request merged**